### ISSUE TITLE
streamline the implementations of iterators to drastically reduce memory allocation

### ISSUE THREAD

**Issue by lemire** [RoaringBitmap/RoaringBitmap#134]
Currently, iterators are implemented by instantiating sub-iterators that run over the containers. This approach requires us to create many more temporary objects than necessary.

A much simpler approach is the one taken by the ``CRoaring`` library where no memory allocation is done... 
https://github.com/RoaringBitmap/CRoaring/blob/master/src/roaring.c#L1156

This should reduce the amount of code *and* reduce the pressure on the garbage collector, which should improve performance in real-life systems.

This is related to https://github.com/RoaringBitmap/RoaringBitmap/issues/44 but distinct.

**Comment by lemire** [comment]
See the following commit for an application of this idea to the ``contains`` function: 
https://github.com/RoaringBitmap/RoaringBitmap/commit/a778e33adb9c651c8c39e60b1eba17a8edc4b0a0

**Comment by ppiotrow** [comment]
What is the reason of having `RoaringIntIterator` https://github.com/RoaringBitmap/RoaringBitmap/blob/master/src/main/java/org/roaringbitmap/RoaringBitmap.java#L49  and `Iterator<Integer>` https://github.com/RoaringBitmap/RoaringBitmap/blob/master/src/main/java/org/roaringbitmap/RoaringBitmap.java#L1442 ? Performance of auto boxing? Is their logic different in some way? I understand that it is not possible to have `Iterator<int>`. 

Second question about `PeekableIntIterator` https://github.com/RoaringBitmap/RoaringBitmap/blob/master/src/main/java/org/roaringbitmap/PeekableIntIterator.java Why are those methods important in real word use cases?

**Comment by lemire** [comment]
@ppiotrow 

Yes, we expect to get better performance with ``IntIterator`` than ``Iterator<Integer>``. It seems that we do too...

        cd RoaringBitmap
        cd jmh
        mvn -f ../pom.xml clean install -DskipTests -Dgpg.skip=true
        mvn -f ../real-roaring-dataset/pom.xml clean install
        mvn  clean install  -DskipTests
        java -jar ./target/benchmarks.jar true -wi 5 -i 5 -f1 IteratorsBenchmark

        IteratorsBenchmark.testBoxed_a             avgt    5     1132.022 ±   1822.187  us/op
        IteratorsBenchmark.testFlyweight_a         avgt    5      651.920 ±      0.169  us/op
        IteratorsBenchmark.testStandard_a          avgt    5      777.727 ±    706.186  us/op

So there are classes where ``Iterator<Integer>`` has almost half the performance in our current code.

**Comment by lemire** [comment]
@ppiotrow 

Please revisit https://github.com/RoaringBitmap/RoaringBitmap/blob/master/src/main/java/org/roaringbitmap/PeekableIntIterator.java I just entered more details in the comments as to what this is good for. The idea is to support more advanced operations efficiently.

**Comment by ppiotrow** [comment]
I think I can port this C code into Java. Currently there multiple iterator implementations in Java, among which there are `(Reverse)IntIteratorFlyweight`. I will try to replace Flyweight iterator with ported C code as their purpose is the same - reduced allocations. Then I'll try to use new implementation as a backend for other iterators to keep iterating code duplication low.

**Comment by lemire** [comment]
@ppiotrow  I have improved the iterator performance in C this week. In particular, the iteration over run containers wasn't great. It is now much better. I was able to double to performance of the C iterators.

My guess is that we can probably double the performance of the Java iterators in some cases by trying the same approach in Java. 

That would be a huge gain.

**Comment by ppiotrow** [comment]
I migrated the C code and wrote unit tests checking correctness on bitmaps generated by `ClusteredDataGenerator`. It detected bug caused by such usage of shift operator https://github.com/RoaringBitmap/CRoaring/blob/master/src/roaring.c#L1178

I wasn't able to fix it and there are two possibilities:
1) I can commit ported code with "minimal" unit test to reproduce and someone more experienced will solve it.
2) I can replace failing operation with approach used currently in Java implementation
https://github.com/RoaringBitmap/RoaringBitmap/blob/master/src/main/java/org/roaringbitmap/BitmapContainer.java#L1348-L1357
The question is why C and Java iteration of `BitmapContainer` differs that much?

**Comment by ppiotrow** [comment]
Found the bug, my `Java` replacement for `UINT64_MAX` was `-1` instead of `-1L`. The question 2) still remains valid. Why are BitmapContainers iterations implementations different in Java and C?

**Comment by lemire** [comment]
@ppiotrow 

<em>I can replace failing operation with approach used currently in Java implementation</em>

I think that it would be fine. I don't think your goal should be to replace the well tested code we have right now in Java, but rather to reorganize it so that we do not create ``ShortIterator`` objects. I suspect that these are a performance problem.

<em>Why are BitmapContainers iterations implementations different in Java and C?</em>

There is an API difference. The Java API has two function calls (``hasNext``) and (``next``). By design, the ``next`` function must not only return the next value, but also gather enough information to determine whether there is a next value after the next value. The C API has a single function call (there is no ``hasNext``)  that populates a structure with both the value (if available) and a flag to indicate whether the value is available.

In C we iterate this way...

```
roaring_uint32_iterator_t   i;
roaring_create_iterator(ra, &i);
while(i.has_value) {
  printf("value = %d\n", i.current_value);
  roaring_advance_uint32_iterator(&i);
}
```

In Java, we do it this way...

```
while(i.hasNext()) {
  int x = i.next() 
}
```

It looks the same, but it is quite different.


For future reference...

Here is the C code...

https://github.com/RoaringBitmap/CRoaring/blob/master/src/roaring.c#L1176-L1188

and here is the Java code...

https://github.com/RoaringBitmap/RoaringBitmap/blob/master/src/main/java/org/roaringbitmap/BitmapContainer.java#L1348-L1357


This is related to an issue raised by @leventov  (see https://github.com/RoaringBitmap/RoaringBitmap/issues/140)
